`ilab` command redesign

cdoern opened the pull request:
add the following groups to ilab:

model
config
data
taxonomy

inside of these commands, there are groups commands that relate to the sub-command. For example `ilab model list` lists all models known to the system.

a full list of the commands in this PR is:

model
 - convert
 - download
 - train
 - chat
 - serve
 - evaluate
 
data
 - generate

config
 - init

taxonomy
 - diff

the goal of this is to make `ilab` commands more logical. `ilab generate` for example is vague. also, adding structure allows for natural expansion of the project and makes the repo more digestible to new contributors.

Grouping ilab commands under common resources which the cli deals with makes sense as the role of instructlab grows. This is the first step in making `ilab` a model engine in which the cli will be a tool for full blown LLM management and deployment

hickeyma reviewed (commented):
This PR should be driven by [design doc](https://github.com/instructlab/dev-docs/pull/48) and not be merged until the design merges first.

hickeyma commented:
Add hold waiting on [design doc](https://github.com/instructlab/dev-docs/pull/48)

hickeyma commented:
Removed hold as [design doc](https://github.com/instructlab/dev-docs/pull/48) was merged.

hickeyma commented:
https://github.com/Mergifyio rebase

russellb commented on src/instructlab/model/chat.py:
```diff
@@ -0,0 +1,771 @@
+# SPDX-License-Identifier: Apache-2.0
```
When you break up a PR into a series of commits, each commit should make sense on its own as a contained step in a progression. This commit adds a ton of code, but doesn't remove any or change any, which doesn't really make sense.

I would like to see a chunk moved in one go. As it stands, I don't think the way it's broken up is any easier to review than the whole thing in a single commit. It still has to be viewed all together to make sense of the moves.

Does that make sense?

russellb replied:
Another suggestion -- it's often much easier to submit refactorings completely separate from functional changes. It's OK if that's separate commits in a single PR. That would help a lot here. It seems you wanted to do a significant reorganization of the code as part of this. It's best to do that with zero functional changes if possible, and then put the functional changes in their own commits later.

russellb replied:
I'm trying to dig up some examples of PRs that have a series of commits that can be reviewed independently. Here are a few ...

- https://github.com/instructlab/instructlab-bot/pull/125/commits
- https://github.com/instructlab/instructlab/pull/994/commits
- https://github.com/instructlab/instructlab/pull/951/commits

or some other resources on the topic, linked to specific relevant sections ...

- https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#smaller-is-better-small-commits-small-pull-requests
- https://www.berrange.com/tags/commit-message/
- https://docs.kernel.org/process/submitting-patches.html#separate-your-changes
-

cdoern commented:
ok @russellb I think this fits more of the norm you are looking for, if not, lets sync tomorrow to see how we can get this all sorted before reviews.

nathan-weinberg commented on src/instructlab/data/generate.py:
```diff
@@ -0,0 +1,218 @@
+
```
Remove this top line

nathan-weinberg commented on src/instructlab/data/generate.py:
```diff
@@ -0,0 +1,218 @@
+@click.option(
+    "--num-cpus",
+    type=click.INT,
+    help="Number of processes to use.",
```
Number of processes or number of CPUs? Not necessarily the same thing

cdoern replied:
True, this was from before. But looking at the code this seems to actually be the amount of processes to spawn not the # of cpus

nathan-weinberg commented on src/instructlab/data/generate.py:
```diff
@@ -0,0 +1,218 @@
+@click.option(
+    "--chunk-word-count",
+    type=click.INT,
+    help="Number of words to chunk the document",
```
```suggestion
    help="Number of words to chunk the document",
```
This could be better IMO

russellb replied:
This is not new content, though

nathan-weinberg replied:
I wouldn't block on it, but if there's a better way to explain I don't see why not improve the text here - similar to https://github.com/instructlab/instructlab/pull/990/files#r1626438457

But again, not a must-have

cdoern replied:
I agree, no new content. Making this a heavier lift is not what we need

nathan-weinberg commented on src/instructlab/data/generate.py:
```diff
@@ -0,0 +1,218 @@
+@click.option(
+    "--api-key",
+    type=click.STRING,
+    default=helpers.DEFAULT_API_KEY,  # Note: do not expose default API key
+    help="API key for API endpoint. [default: manage.DEFAULT_API_KEY]",
```
Is this right?

nathan-weinberg commented on src/instructlab/data/generate.py:
```diff
@@ -0,0 +1,218 @@
+    default="",
+    help="TLS client certificate password.",
+)
+@click.option(
+    "--model-family",
+    help="Force model family to use when picking a generation template",
+)
```
This seems like it's missing several fields

russellb replied:
which ones?

or is this comment now out of date?

nathan-weinberg replied:
There's just `help` here - we should at least have a `type` I presume

cdoern replied:
this is in the main branch as is

nathan-weinberg commented on src/instructlab/config/init.py:
```diff
@@ -0,0 +1,124 @@
+    cfg.generate.taxonomy_base = taxonomy_base
+    helpers.write_config(cfg)
+
+    click.echo(
+        "Initialization completed successfully, you're ready to start using `ilab`. Enjoy!"
+    )
```
More of a UX suggestion here
```suggestion
    click.echo(
        "Initialization completed successfully, you're ready to start using `ilab`. Enjoy!",
        fg="green"
    )
```

nathan-weinberg commented on src/instructlab/helpers.py:
```diff
@@ -0,0 +1,372 @@
+        for item in dict:
+            configP[item] = dict[item]
+
+       # configP['conf'] = get_dict(cfg)
+        configP.write(conffile)
+        #.safe_dump(, stream=yamlfile)
```
Commented out lines?

nathan-weinberg commented on src/instructlab/model/model.py:
```diff
@@ -0,0 +1,51 @@
+@click.pass_context
+# pylint: disable=redefined-outer-name
+def model(ctx, config_file):
+    #print(ctx.obj)
+    #print(ctx.parent.obj)
+    #ctx = ctx.parent
+    """Command Group for Interacting with the Models in InstructLab.
+
+    If this is your first time running ilab, it's best to start with `ilab init` to create the environment.
+    """
+    helpers.init_config(ctx, config_file)
+   # ctx = ctx.parent
+   # pass
```
Commented out lines?

nathan-weinberg commented on src/instructlab/model/serve.py:
```diff
@@ -0,0 +1,312 @@
+@click.pass_context
+def serve(ctx, model_path, gpu_layers, num_threads, max_ctx_size, model_family):
+    """Start a local server"""
+    # pylint: disable=C0415
+    # Local
+
+
```
Remove this if there are no local imports

nathan-weinberg commented on src/instructlab/taxonomy/diff.py:
```diff
@@ -0,0 +1,83 @@
+        )
+
+
+# ilab list => ilab diff
+# ilab check => ilab diff --quiet
+#utils.make_lab_diff_aliases(ilab, diff)
```
Can remove these

russellb replied:
please remove the commented out code

nathan-weinberg commented on src/instructlab/taxonomy/diff.py:
```diff
@@ -0,0 +1,83 @@
+        raise SystemExit(1) from exc
+    if not quiet:
+        click.secho(
+            f"Taxonomy in /{taxonomy_path}/ is valid :)",
```
```suggestion
            f"Taxonomy in /{taxonomy_path}/ is valid :)",
```
I've never understood why the slashes around `{taxonomy_path}` were added

cdoern replied:
removed, I agree

cdoern commented:
working on getting the alias commands working tonight. click does not have a nice way to do this through different parent cmd groups

russellb commented on src/instructlab/helpers.py:
```diff
@@ -0,0 +1,369 @@
+# SPDX-License-Identifier: Apache-2.0
```
Can you help me understand why these changes are necessary in this PR? 

The commit seems to move the config code, but not remove the old copy.  Also, none of the history from either of the combined files is retained since git does not see any of it as a move.

cdoern replied:
each time I rebase, git (or me) seems to forget config.py. I will fix it on the next rebase or I can redo it

russellb commented on pyproject.toml:
```diff
@@ -32,7 +32,7 @@ dynamic = ["dependencies", "optional-dependencies", "version"]
 [project.scripts]
 # defines ilab executable
-ilab = "instructlab.lab:cli"
+ilab = "instructlab.lab:ilab"
```
This commit is nicely isolated to show one change. Nice work

cdoern commented:
This is almost ready. One more rebase.

cdoern commented on src/instructlab/model/serve.py:
```diff
@@ -181,7 +241,7 @@ def read_root():
     eos_token = "<|endoftext|>"
     bos_token = ""
     for template_dict in templates:
-        if template_dict["model"] == get_model_family(model_family, model_path):
```
TODO for me in the morning, fix this

nathan-weinberg replied:
How'd it go? :wink:

russellb commented on src/instructlab/data/data.py:
```diff
@@ -0,0 +1,26 @@
+
+@click.group(chain=True)
+@click.version_option(package_name="cli")
+@click.option(
+    "--config",
+    "config_file",
+    type=click.Path(),
+    default=helpers.DEFAULT_CONFIG,
+    show_default=True,
+    help="Path to a configuration file.",
+)
```
Similar to my comment on the `model` group, we know have `--config` in 3 places instead of just using the top level option. I'd prefer we just keep the top-level one.

cdoern replied:
this was a necessary change not one I really wanted to do. Click has no way of passing the ctx multiple levels, so the ctx with the .obj needs to be initialized on the resource level

russellb commented on src/instructlab/data/generator/generate_data.py:
```diff
@@ -20,13 +20,9 @@
 import tqdm
```
It would be helpful to not move any of the code under the `generator` directory. All of that is moving to the `instructlab.sdg` library.

Ideally it wouldn't be changed, either, other than what is absolutely necessary.

cdoern replied:
For now, I think the sake of this design generator should live in `data` when sdg library comes around this dir can simply be deleted hopefully.

russellb commented on src/instructlab/data/generate.py:
```diff
@@ -0,0 +1,218 @@
+    default=helpers.DEFAULT_API_KEY,  # Note: do not expose default API key
+    help="API key for API endpoint. [default: manage.DEFAULT_API_KEY]",
+)
+@click.option(
+    "--yaml-rules",
+    type=click.Path(),
+    default=helpers.DEFAULT_YAML_RULES,
+    show_default=True,
+    help="Rules file for YAML linting",
+)
```
you've got a default here that wasn't in the original code

nathan-weinberg replied:
It wasn't here exactly, but it was nested in the `utils.py` file: https://github.com/instructlab/instructlab/blob/main/src/instructlab/utils.py#L28

@cdoern is there a design change accounting for this move? If the functionality is the same I think it's fine to have here

cdoern replied:
fixed (I think?) in my undoing of helpers.py

russellb commented on src/instructlab/data/generate.py:
```diff
@@ -0,0 +1,218 @@
+        click.echo(
+            f"Generating synthetic data using '{model}' model, taxonomy:'{taxonomy_path}' against {api_base} server"
+        )
+        generate_data(
+            logger=logger,
+            api_base=api_base,
+            api_key=api_key,
+            model_name=model,
+            num_cpus=num_cpus,
+            num_instructions_to_generate=num_instructions,
+            taxonomy=taxonomy_path,
+            taxonomy_base=taxonomy_base,
+            output_dir=output_dir,
+            prompt_file_path=prompt_file_path,
+            rouge_threshold=rouge_threshold,
+            console_output=not quiet,
+            yaml_rules=yaml_rules,
+            chunk_word_count=chunk_word_count,
+            server_ctx_size=server_ctx_size,
+        )
```
This appears to have dropped TLS support from generate

cdoern replied:
fixed

russellb commented on src/instructlab/generator/README.md:
```diff
@@ -1,207 +0,0 @@
-### Notice
```
It looks like you dropped this file completely

nathan-weinberg replied:
Seems this is a LICENSE, not an actual README - can it be dropped?

russellb commented on src/instructlab/config/config.py:
```diff
@@ -0,0 +1,15 @@
+#from lab import ilab
```
Please remove the commented out line and add the missing license header

russellb commented on src/instructlab/config/init.py:
```diff
@@ -0,0 +1,124 @@
+import click
```
add missing license header

russellb commented on src/instructlab/config/config.py:
```diff
@@ -0,0 +1,15 @@
+import click
+from .init import init
+
+@click.group(chain=True)
```
I don't think `chain=True` is appropriate here, but let me know if I'm missing something

cdoern replied:
lets sync about `chain=True` click doesn't like when this option isn't specified (click weirdness we need to move away from in the future!)

russellb replied:
Can you be a bit more explicit about the click weirdness?

russellb commented on src/instructlab/config/init.py:
```diff
@@ -0,0 +1,124 @@
+        )
+
+    try:
+        taxonomy_contents = listdir(taxonomy_path)
```
I would prefer this line was left as-is -- `os.listdir()` instead of `listdir()`

russellb replied:
mostly to minimize unnecessary code change

... secondly a matter of taste. I'm not a big fan of importing everything into the base namespace. I like seeing where something is coming from without having to go back to the top to figure it out.

russellb commented on src/instructlab/taxonomy/diff.py:
```diff
@@ -0,0 +1,83 @@
+import click
```
add missing license header

russellb commented on src/instructlab/taxonomy/taxonomy.py:
```diff
@@ -0,0 +1,27 @@
+#from lab import ilab
```
add missing license header

russellb commented on src/instructlab/taxonomy/taxonomy.py:
```diff
@@ -0,0 +1,27 @@
+from .diff import diff
+from instructlab import helpers
+
+@click.group(chain=True)
```
same comment about `chain=True` here

cdoern replied:
We can discuss this in a thread, but I am pretty sure chain=True is needed? Click gets confused without it in my testing. Though I can try again

alimaredia replied:
@russellb @cdoern Can all of these `chain=True` comments be resolved/dismissed if they aren't still relevant? There are many across many files, and it's not clear since the comments are so old if the comments are still relevant or not.

russellb commented on src/instructlab/taxonomy/taxonomy.py:
```diff
@@ -0,0 +1,27 @@
+@click.group(chain=True)
+@click.version_option(package_name="cli")
+@click.pass_context
+@click.option(
+    "--config",
+    "config_file",
+    type=click.Path(),
+    default=helpers.DEFAULT_CONFIG,
+    show_default=True,
+    help="Path to a configuration file.",
+)
```
same comment about the duplicated `--config` option here

cdoern replied:
see my other comments about this, kind of a dealbreaker

russellb commented on requirements.txt:
```diff
@@ -40,3 +40,4 @@ langchain-text-splitters
 # the below library should NOT be imported into any python files
 # it is for CLI usage ONLY
 yamllint>=1.35.1,<1.36.0
+prettytable
```
It would be better to include this in the change that makes use of the dependency.

Also, it should have version guardrails similar to the rest of the entries.

cdoern replied:
```ERROR: Ignored the following versions that require a different python version: 1.21.2 Requires-Python >=3.7,<3.11; 1.21.3 Requires-Python >=3.7,<3.11; 1.21.4 Requires-Python >=3.7,<3.11; 1.21.5 Requires-Python >=3.7,<3.11; 1.21.6 Requires-Python >=3.7,<3.11
ERROR: Could not find a version that satisfies the requirement prettytable<4.0.0,>=3.10.1 (from versions: 0.3, 0.4, 0.5, 0.6, 0.6.1, 0.7.1, 0.7.2, 1.0.0, 1.0.1, 2.0.0, 2.1.0, 2.2.0, 2.2.1, 2.3.0, 2.4.0, 2.5.0, 3.0.0, 3.1.0, 3.1.1, 3.2.0, 3.3.0, 3.4.0, 3.4.1, 3.5.0, 3.6.0, 3.7.0, 3.8.0, 3.9.0, 3.10.0)```

I am going to set ==3.10.0 or now unless someone says otherwise

russellb commented on src/instructlab/__main__.py:
```diff
@@ -3,4 +3,4 @@
 # First Party
 from instructlab import lab
 
-lab.cli(None, None)
```
I'm guessing this was supposed to be part of the 2nd commit in your series?

cdoern replied:
fixed, the switch from `.cli` to `.ilab` should happen in the early commit but the args change later.

russellb commented on src/instructlab/data/generate.py:
```diff
@@ -11,7 +10,7 @@
 )
 @click.option(
-    "--num-cpus",
+    "--num-processes",
```
This is a functional change that's hidden in a cleanup commit. It should really be surfaced on its own. It's also breaking backwards compatibility where your design otherwise goes out of its way to ensure backwards compatibility.

cdoern replied:
True, I will leave as cpu for now

russellb commented on src/instructlab/config/init.py:
```diff
@@ -111,14 +110,14 @@ def init(
-    cfg.chat.model = model
-    cfg.generate.model = model
+    cfg.chat.model = model_path
+    cfg.generate.model = model_path
```
It looks like this belonged in an earlier commit?

cdoern replied:
removed

russellb commented on src/instructlab/helpers.py:
```diff
@@ -37,8 +37,9 @@
-DEFAULT_MODEL_PATH = f"models/{DEFAULT_MODEL}.gguf"
+DEFAULT_MODEL_OLD = "merlinite-7b-lab-Q4_K_M"
+DEFAULT_MODEL = "models/merlinite-7b-lab-Q4_K_M.gguf"
+DEFAULT_MODEL_PATH = "models/merlinite-7b-lab-Q4_K_M.gguf"
```
I think I'll stop commenting on this commit at this point with a more general comment --

it looks like the commit is mostly changes that needed to be incorporated in earlier steps in the process. I spent a LOT of time going through and identifying places where code changes were missed in past commits and I see some of them were addressed here.

russellb commented on src/instructlab/lab.py:
```diff
@@ -54,16 +60,15 @@
+ilab.add_command(data_group.data)
+ilab.add_command(config_group.config)
+
+
```
one too many blank lines here

russellb commented on src/instructlab/config/config.py:
```diff
@@ -1,7 +1,11 @@
 import click
+
+# Local
 from .init import init
```
Is this commit mostly changes that the linter/formatter pointed out and you included the fixes here?

cdoern replied:
i moved most (if not all) of the linting into the proper commits

russellb commented on src/instructlab/lab.py:
```diff
@@ -4,7 +4,6 @@
 import typing
 
 # Third Party
-from click_didyoumean import DYMGroup
```
I think the `click_didyoumean` functionality is worth keeping.

I found a way to do aliasing with a much smaller change and without removing `DYMGroup` in #1264.

For reference for anyone else looking ... https://github.com/click-contrib/click-didyoumean

cdoern replied:
dym is still in the resource groups which is where the cmds actually are.

I can add something like dym to my custom class :shrug:

russellb replied:
I think it's at least worth a follow-up issue to re-add the UX polish provided by didyoumean to the top-level

russellb replied:
https://github.com/instructlab/instructlab/issues/1325

russellb commented on src/instructlab/log.py:
```diff
@@ -37,7 +37,7 @@ def isatty(self):
 
 
-def stdout_stderr_to_logger(logger, log_file):
+def stdout_stderr_to_logger(logger, log_file=None):
```
I believe you made this change because a previous commit accidentally dropped some log file support that was added

nathan-weinberg replied:
@leseb IIRC you added that support yes? Apologies for the ping if I am wrong

cdoern replied:
will remove this, I think I fixed the broken commits that caused this to error out

russellb commented on src/instructlab/data/generator/utils.py:
```diff
@@ -17,7 +17,7 @@
 # First Party
 from instructlab.helpers import DEFAULT_API_KEY, DEFAULT_MODEL_OLD
-from instructlab.utils import get_sysprompt, get_ssl_cert_config
+from instructlab.utils import get_ssl_cert_config, get_sysprompt
```
It would be nicer and easier to review for all lining fixes like this to be done along the way within each commit

russellb reviewed (changes requested):
I completed a review pass

nathan-weinberg commented on src/instructlab/data/generate.py:
```diff
@@ -0,0 +1,218 @@
+
```
+1

nathan-weinberg commented on src/instructlab/data/generate.py:
```diff
@@ -0,0 +1,224 @@
+            # TODO: check for working offloading. The function only checks
+            # for compile time defines like `GGML_USE_CUDA`.
+            click.secho(
+                "llama_cpp_python is built without hardware acceleration. "
```
nit
```suggestion
                "llama_cpp_python is built without hardware acceleration."
```

xukai92 replied:
+1. the space should be there as it concatenates the next line

nathan-weinberg commented on src/instructlab/helpers.py:
```diff
@@ -14,19 +19,30 @@
 DEFAULT_MODEL = "models/merlinite-7b-lab-Q4_K_M.gguf"
 DEFAULT_MODEL_PATH = "models/merlinite-7b-lab-Q4_K_M.gguf"
-DEFAULT_TAXONOMY_REPO = "https://github.com/instructlab/taxonomy.git"
+DEFAULT_TAXONOMY_REPO = "git@github.com:instruct-lab/taxonomy.git"
```
Why the change here from `https` to `ssh`?

nathan-weinberg commented on src/instructlab/helpers.py:
```diff
@@ -14,19 +19,30 @@
+HOME = os.getenv("HOME")
 DEFAULT_API_KEY = "no_api_key"
 DEFAULT_CONFIG = "config.yaml"
-# TODO: Consolidate --model and --model-path into one --model-path flag since we always need a path now
 DEFAULT_MODEL_OLD = "merlinite-7b-lab-Q4_K_M"
 DEFAULT_MODEL = "models/merlinite-7b-lab-Q4_K_M.gguf"
 DEFAULT_MODEL_PATH = "models/merlinite-7b-lab-Q4_K_M.gguf"
-DEFAULT_TAXONOMY_REPO = "https://github.com/instructlab/taxonomy.git"
+DEFAULT_TAXONOMY_REPO = "git@github.com:instruct-lab/taxonomy.git"
 DEFAULT_TAXONOMY_PATH = "taxonomy"
 DEFAULT_TAXONOMY_BASE = "origin/main"
-DEFAULT_YAML_RULES = "yaml_rules.yaml"
+DEFAULT_YAML_RULES = """\
+extends: relaxed
+
+rules:
+  line-length:
+    max: 120
+"""
```
Think this change related to the convo happening further up

nathan-weinberg commented on src/instructlab/helpers.py:
```diff
@@ -161,6 +216,7 @@ def get_default_config():
 
 def read_config(config_file=DEFAULT_CONFIG):
     """Reads configuration from disk."""
+    # pylint: disable=W0108
```
:eyes:

nathan-weinberg commented on src/instructlab/helpers.py:
```diff
@@ -197,4 +253,55 @@ def get_api_base(host_port):
+        elif not os.path.isfile(config_file):
+            config_obj = None
+            ctx.fail(
+                f"`{config_file}` does not exists, please run `ilab init` "
```
```suggestion
                f"`{config_file}` does not exists, please run `ilab config init` "
```

nathan-weinberg commented on src/instructlab/model/chat.py:
```diff
@@ -55,6 +67,211 @@
+@click.option(
+    "--endpoint-url",
+    type=click.STRING,
+    help="Custom URL endpoint for OpenAI-compatible API. Defaults to the `ilab serve` endpoint.",
```
Should this be `ilab model serve`?

nathan-weinberg commented on src/instructlab/model/chat.py:
```diff
@@ -55,6 +67,211 @@
+@click.option(
+    "--api-key",
+    type=click.STRING,
+    default=DEFAULT_API_KEY,  # Note: do not expose default API key
```
Comment needed?

nathan-weinberg replied:
Also, should this be `helpers.DEFAULT_API_KEY`?

cdoern replied:
comment was there, and I imported the vars directly from `config` so no need for the `.DEFAULT...`

nathan-weinberg commented on src/instructlab/model/chat.py:
```diff
@@ -55,6 +67,211 @@
+    "--api-key",
+    type=click.STRING,
+    default=DEFAULT_API_KEY,  # Note: do not expose default API key
+    help="API key for API endpoint. [default: config.DEFAULT_API_KEY]",
```
This text is no longer accurate

cdoern replied:
switched back to config file

nathan-weinberg commented on src/instructlab/model/chat.py:
```diff
@@ -55,6 +67,211 @@
+    tls_client_passwd,
+    model_family,
+):
+    """Run a chat using the modified model"""
```
Not necessarily a modified model no?

cdoern replied:
This was the help text from before, not looking to change stuff like that in my PR I believe

nathan-weinberg commented on src/instructlab/model/convert.py:
```diff
@@ -0,0 +1,106 @@
+import click
+
+# First Party
+# from instructlab.train.lora_mlx.mlx_explore import gguf_convert_to_mlx
```
Remove commented out import

nathan-weinberg commented on src/instructlab/model/convert.py:
```diff
@@ -0,0 +1,106 @@
+@click.option(
+    "-sd",
+    "--skip-de-quantize",
+    help="Skip de-quantization.",
```
nit: This helptext doesn't really tell me anything

nathan-weinberg commented on src/instructlab/model/evaluate.py:
```diff
@@ -0,0 +1,74 @@
+
+    adapter_file_exists = adapter_file and os.path.exists(adapter_file)
+    if adapter_file and not adapter_file_exists:
+        print(
```
Won't block on this, but why are we using print statements here instead of the logger?

russellb replied:
This was the existing code, so let's address it in a future PR

nathan-weinberg commented on src/instructlab/model/model.py:
```diff
@@ -0,0 +1,53 @@
+
+@click.group(cls=DYMGroup, chain=True)
+@click.version_option(package_name="cli")
+@click.option(
+    "--config",
+    "config_file",
+    type=click.Path(),
+    default=helpers.DEFAULT_CONFIG,
+    show_default=True,
+    help="Path to a configuration file.",
+)
```
Yeah I don't think subcommands should have config overrides other than individual variables via a command line option

cdoern replied:
--config can't be an `ilab` level option anymore as the `init_config` needs to happen on the resource group level due to context passing norms in click. There is no way to pass the full context with the config obj down two levels in click.

russellb replied:
We should look at this again this week. I didn’t move it in my quick version and it was still working. The click structure is a little different, I think.

nathan-weinberg commented on src/instructlab/model/model.py:
```diff
@@ -0,0 +1,52 @@
+def model(ctx, config_file):
+    """Command Group for Interacting with the Models in InstructLab.
+
+    If this is your first time running ilab, it's best to start with `ilab init` to create the environment.
```
Broken record here but think we might want this to be `ilab config init`

russellb replied:
https://github.com/instructlab/instructlab/issues/1331

nathan-weinberg commented on src/instructlab/model/serve.py:
```diff
@@ -15,12 +15,72 @@
+)
+@click.option(
+    "--model-family",
+    type=str,
```
Isn't there a `click.STRING` type?

nathan-weinberg commented on src/instructlab/taxonomy/diff.py:
```diff
@@ -0,0 +1,83 @@
+    type=click.Path(),
+    help=f"Path to {helpers.DEFAULT_TAXONOMY_REPO} clone or local file path.",
+)
+@click.option(
+    "--taxonomy-base",
+    help="Base git-ref to use for taxonomy.",
+)
```
This is different than the other two definitions of this option

cdoern replied:
wdym? this is what was copied from lab.py

alimaredia replied:
As Charlie mentioned, this is just a copy from lab.py as intended. I'm resolving the converation for other reviewers sake.

nathan-weinberg commented on src/instructlab/taxonomy/diff.py:
```diff
@@ -0,0 +1,83 @@
+@click.option(
+    "--yaml-rules",
+    type=click.Path(),
+    default=None,
```
This is different than what you did above

cdoern replied:
copied from lab.py

markstur commented on src/instructlab/config.py:
```diff
@@ -161,6 +168,7 @@ def get_default_config():
 
 def read_config(config_file=DEFAULT_CONFIG):
     """Reads configuration from disk."""
+    # pylint: disable=W0108
```
Where is the unnecessary lambda?   Don't see need to add this pylint disable

cdoern replied:
sorry, left over. Will remove thank for catching

markstur commented on src/instructlab/model/chat.py:
```diff
@@ -55,6 +66,211 @@
+                    else model
+                )
+
+            except config.ClientException as exc:
```
AttributeError: module 'instructlab.config' has no attribute 'ClientException'

markstur commented on src/instructlab/model/chat.py:
```diff
@@ -55,6 +66,211 @@
+            and api_base == ctx.obj.config.serve.api_base()
+        ):
+            try:
+                models = config.list_models(
```
AttributeError: module 'instructlab.config' has no attribute 'list_models'

markstur commented on src/instructlab/model/chat.py:
:shrug:  still not sure chat belongs under model.  New semantics are a little confusing to me

cdoern replied:
`ilab model chat` since you are chatting with a model. The files and cmds are being grouped by resource and the commands that fall under the resource are actions performed on or associated with the resource. 

kinda a new way of thinking about it all, i know, but I think this makes sense!

markstur reviewed (dismissed):
Seeing some AttributeErrors

markstur commented on src/instructlab/lab.py:
```diff
@@ -43,1332 +31,67 @@
+        if self.aliases:
+            formatter.write_paragraph()
+            formatter.write_text("Aliases:")
+            with formatter.indentation():
```
indentaion isn't working.   Would be nice to have indent after Aliases:

```
Aliases:
chat: model chat
convert: model convert
diff: taxonomy diff
download: model download
evaluate: model evaluate
generate: data generate
init: config init
list: model list
serve: model serve
train: model train

```

russellb replied:
https://github.com/instructlab/instructlab/issues/1324

markstur reviewed (commented):
few more inline

cdoern commented on src/instructlab/model/chat.py:
```diff
@@ -55,6 +67,211 @@
+    default="",
+    help="TLS client certificate password.",
+)
+@click.option(
+    "--model-family",
+    default="merlinite",
+    show_default=True,
+    help="model family to use when picking a chat template",
+)
```
fixed, thanks

cdoern commented on .gitignore:
```diff
@@ -2,7 +2,7 @@
 !src/instructlab/train/lora_mlx/models
 # We recommend using taxonomy here, but not a submodule. Just a work dir.
-taxonomy/
+./taxonomy/
```
hm, we need something. or I guess I can call the new directory `tax`. If its called taxonomy and this rule is in place, the the files changed there do not get added

cdoern commented on src/instructlab/model/train.py:
```diff
@@ -0,0 +1,363 @@
+    show_default=True,
+    help="model name to use in training",
+)
+@click.pass_context
+def train(
+    ctx,
+    data_dir,
+    input_dir,
+    skip_preprocessing,
+    tokenizer_dir,
+    gguf_model_path,
+    model_dir,
+    iters,
+    local,
+    skip_quantize,
+    num_epochs,
+    device: "torch.device",
+    four_bit_quant: bool,
+    model_name: str,
+):
```
fixed in most recent commits, thank you for catching this.

cdoern commented on src/instructlab/model/train.py:
```diff
@@ -0,0 +1,363 @@
+
+    name = "deviceinfo"
+    supported_devices = {"cuda", "cpu"}
+
+    def convert(self, value, param, ctx) -> "torch.device":
+        # pylint: disable=C0415
+        # Function local import, import torch can take more than a second
+        # Third Party
+        import torch
+
+        if not isinstance(value, torch.device):
+            try:
+                device = torch.device(value)
+            except RuntimeError as e:
+                self.fail(str(e), param, ctx)
+
+        if device.type not in {"cuda", "cpu"}:
+            supported = ", ".join(repr(s) for s in sorted(self.supported_devices))
+            self.fail(
+                f"Unsupported device type '{device.type}'. Only devices "
+                f"types {supported}, and indexed device strings like 'cuda:0' "
+                "are supported for now.",
+                param,
+                ctx,
+            )
+
+        # Detect CUDA/ROCm device
+        if device.type == "cuda":
+            if not torch.cuda.is_available():
+                self.fail(
+                    f"{value}: Torch has no CUDA/ROCm support or could not detect "
+                    "a compatible device.",
+                    param,
+                    ctx,
+                )
+            # map unqualified 'cuda' to current device
+            if device.index is None:
+                device = torch.device(device.type, torch.cuda.current_device())
+
+        return device
```
fixed, thanks!

cdoern commented on src/instructlab/model/convert.py:
```diff
@@ -0,0 +1,101 @@
+    if not utils.is_macos_with_m_chip():
+        convert_llama_to_gguf(model=model_dir, pad_vocab=True)
+    else:
+        from instructlab_quantize import run_quantize  # pylint: disable=import-error
```
undid these changes, will table them for a future PR

cdoern commented:
@russellb I did a lot of re-organization of commits on this PR on fri, sat, sun and I think its in a pretty good spot. I went through your review primarily. I still need to go over Nathan's in detail but I think most of his larger comments overlapped with yours. 

I'd like to address two things

1. the movement of --config

--config needs to be on the resource level, why? This is because click does not have a way to initialize a context object and pass it, with its objects in tact to multiple levels of commands. This means we lose the `Lab` object with the config inside and many of the benefits of the ctx such as auto-populating many arguments from the cfg.

2. DYM class

the DYM group still exists on the resource level (where it matters) but on JUST the ilab command we now use an alias custom class. IMO this isn't a huge deal since the cmds which actually may be mistyped still have the dym.

Everything else was organizational. While I'd love to rebase this PR all week (kidding) I think this structure is good enough and we should focus on getting this in early this week.

The release is scheduled for Friday, if this is not in before weds I think we should discuss moving the release to the following Monday to give it time to simmer. This PR should be viewed as a blocker for the release I think. 

lmk what you think!

leseb commented on src/instructlab/data/data.py:
```diff
@@ -0,0 +1,33 @@
+
+
+@click.group(chain=True)
+@click.version_option(package_name="cli")
```
Did you mean `instructlab`?

cdoern replied:
thanks!

leseb commented on src/instructlab/configuration/config.py:
```diff
@@ -0,0 +1,19 @@
+def config():
+    """Command Group for Interacting with the Config of InstructLab.
+
+    If this is your first time running ilab, it's best to start with `ilab init` to create the environment.
```
Even though `ilab init` still works (right?), I think we should print `ilab config init` instead.

alimaredia replied:
This message has been repeated in `taxonomy/taxonomy.py` and `data/data.py`

russellb replied:
https://github.com/instructlab/instructlab/issues/1331

leseb commented on src/instructlab/tax/diff.py:
```diff
@@ -0,0 +1,90 @@
+# SPDX-License-Identifier: Apache-2.0
```
I don't mind having: `taxonomy` instead of  `tax` directory. We don't have any  `tax` command either. So I'd like to keep the entire word.

cdoern replied:
having `taxonomy` as the name breaks the gitignore. From what I can see (I may be wrong) there is no way to exclude `/taxonomy/` and not `/src/instructlab/taxonomy`. So I named it tax

leseb replied:
Would this solve your problem?

```
diff --git a/.gitignore b/.gitignore
index b650cec..1f492bd 100644
--- a/.gitignore
+++ b/.gitignore
@@ -3,6 +3,7 @@ models/
 !src/instructlab/train/lora_mlx/models
 # We recommend using taxonomy here, but not a submodule. Just a work dir.
 taxonomy/
+!src/instructlab/taxonomy/
 # We recommend putting generated files here, but PLEASE don't commit them.
 generated/
 # We recommend outputting training results on Linux here, but PLEASE don't commit them.
```

Then:

```
(venv) 👺  instructlab - [main] ❯❯❯ mkdir -p src/instructlab/taxonomy/ && touch src/instructlab/taxonomy/foo.py                  ✹
(venv) 👺  instructlab - [main] ❯❯❯ git status                                                                                 ✹ ✭
On branch main
Your branch is ahead of 'origin/main' by 453 commits.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   .gitignore

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        src/instructlab/taxonomy/

no changes added to commit (use "git add" and/or "git commit -a")
```

russellb replied:
agree -- would prefer the `taxonomy` dir to match the command name (and the gitignore fix given by leseb)

xukai92 replied:
+1 let's keep `taxonomy`

cdoern replied:
fixed!

leseb commented on src/instructlab/tax/taxonomy.py:
```diff
@@ -0,0 +1,33 @@
+@click.group(chain=True)
+@click.version_option(package_name="cli")
+@click.option(
+    "--config",
```
It makes me sad to see that each time with add a group we have to re-add this, isn't there any way not to rewrite that part each time?

cdoern replied:
click weirdness, this flag needs to be at the resource level in order to pass the context down properly

leseb commented on src/instructlab/tax/taxonomy.py:
```diff
@@ -0,0 +1,33 @@
+    help="Path to a configuration file.",
+)
+@click.pass_context
+# pylint: disable=redefined-outer-name
```
Why do we need this? Where is the `redefined-outer-name`?

cdoern replied:
removed

leseb commented on src/instructlab/lab.py:
```diff
@@ -36,7 +35,51 @@
 
 
-@click.group(cls=DYMGroup)
+class ExpandAliasesGroup(click.Group):
```
It was discussed during one of the community meetings that when users input an "old" command, they should get a warning that the command will be deprecated in 2 releases (I'm making this up, will need to agree as a group). I don't see it happening here. Did I miss it?

cdoern replied:
added

leseb reviewed (dismissed):
This is a large change, but I'm generally happy with the direction it takes us. I've focused my review on the new and reorganized code rather than fixing existing issues, even though it’s tempting. I believe all maintainers should do the same; otherwise, this change will grow even larger and take longer to review, delaying the merge further.

As I review the code, it feels like this should have been split into multiple PRs. The first PR could have included the initial two commits along with the integration and aliases, establishing a solid foundation for future additions. This would have made the refactor easier to review and track.

I also think it would be beneficial to have an integration test that validates the new commands. If the current tests pass, it indicates backward compatibility, but we currently have no validation for the new commands. Additionally, the e2e test hasn't been updated with the new CLI. Perhaps this is on your list of follow-ups, but before we merge, we should identify all outstanding items. I'm okay with leaving a few items like the alias warnings, but they should be tracked as issues.

Thanks again for your hard work and dedication on this one :D.

EDIT: please update the `CHANGELOG.md`.

cdoern commented on src/instructlab/model/list.py:
```diff
@@ -0,0 +1,42 @@
+    default="models",
+    show_default=True,
+)
+def list(model_dir: str, *args, **kwargs):
```
I feel pretty strongly about having this as we are exposing `models` as a resource group. If there are no obvious issues with the code, I don't see why this is a huge lift

russellb replied:
Please submit it as a separate PR as a matter of separating your refactor from distinct new features

xukai92 replied:
+1 this should be a separate PR

alimaredia replied:
I'm going to have to agree here. Anything to slim down this PR, even at the margins helps.

xukai92 replied:
ps: this could nicely handle checkpoint listing

cdoern commented on src/instructlab/config.py:
```diff
@@ -216,3 +223,50 @@ def get_model_family(forced, model_path):
+
+    
+
+def init_config(ctx, config_file):
```
is this a blocker to merge this PR? I think init_config is fine and more descriptive IMO and this would create an unnecessary rebase on all the commits.

xukai92 replied:
i don't think this is a blocker but it would be good to create an issue for this.
having `config.init_config()` does indicate something off with the abstraction.

russellb replied:
https://github.com/instructlab/instructlab/issues/1323

cdoern commented on src/instructlab/config/init.py:
```diff
@@ -0,0 +1,127 @@
+    "--repository",
+    default=helpers.DEFAULT_TAXONOMY_REPO,
+    show_default=True,
+    help="Taxonomy repository location.",
```
changing this would be a functional change, leaving what is in main

cdoern commented on src/instructlab/data/generate.py:
```diff
@@ -0,0 +1,224 @@
+@click.option(
+    "--api-key",
+    type=click.STRING,
+    default=helpers.DEFAULT_API_KEY,  # Note: do not expose default API key
```
leaving what is in main branch here

cdoern commented on src/instructlab/data/generate.py:
```diff
@@ -0,0 +1,224 @@
+            api_base = ctx.obj.config.serve.api_base()
+    try:
+        click.echo(
+            f"Generating synthetic data using '{model}' model, taxonomy:'{taxonomy_path}' against {api_base} server"
```
leaving what is in main branch for error messaging tests

russellb commented on src/instructlab/configuration/init.py:
```diff
@@ -0,0 +1,129 @@
+            if not overwrite:
+                return
+        click.echo(
+            "Welcome to InstructLab CLI. This guide will help you to setup your environment."
```
You changed this line in an early commit, and then changed it again here. It'd be nice to drop the earlier change in the earlier commit.

not critical though ... just if you're in there anyway ...

russellb commented on src/instructlab/configuration/init.py:
```diff
@@ -0,0 +1,129 @@
+        )
+
+    try:
+        taxonomy_contents = listdir(taxonomy_path)
```
I would prefer this remained `os.listdir()` to reduce the diff when making sure accidental changes weren't included or newer changes lost by accident

(I know this was in an older comment, so apologies if this is a duplicate. I don't know where the old comment went)

russellb commented on src/instructlab/configuration/init.py:
```diff
@@ -0,0 +1,129 @@
+
+    click.secho(
+        "Initialization completed successfully, you're ready to start using `ilab`. Enjoy!",
+        fg="green",
```
I guess you added the color here?

I know it's tempting to make these improvements along the way, but they do make the review a little more difficult.

The change here seems OK, so I won't block on fixing it,  but just FYI for the future -- try to include any functional changes of any kind in a separate commit (and maybe PR, depending on the size of the change)

alimaredia replied:
Agree!

russellb commented on src/instructlab/tax/diff.py:
```diff
@@ -0,0 +1,90 @@
+
+# ilab list => ilab diff
+# ilab check => ilab diff --quiet
+# utils.make_lab_diff_aliases(ilab, diff)
```
This was uncommented in the old code but is commented out here. Is it needed or not?  Either uncomment or remove

russellb commented on src/instructlab/lab.py:
```diff
@@ -3,33 +3,31 @@
 from .sysinfo import get_sysinfo
+from .tax import taxonomy as taxonomy_group
 
-# 'fork' is unsafe and incompatible with some hardware accelerators.
-# Python 3.14 will switch to 'spawn' on all platforms.
```
This shouldn't be removed

cdoern replied:
added back

russellb commented on src/instructlab/lab.py:
```diff
@@ -39,6 +37,7 @@
 
 
 @click.group(cls=DYMGroup)
+@click.version_option(package_name="instructlab")
```
I'm curious why this line was moved? Was it required?  Or just leftover from past moves?

cdoern replied:
no specific reason, I will put back, sorry!

russellb commented on src/instructlab/lab.py:
```diff
@@ -47,20 +46,22 @@
     """
-    # ilab init or "--help" have no config file. ilab sysinfo does not need one.
-    # CliRunner does fill ctx.invoke_subcommand in option callbacks. We have
-    # to validate config_file here.
```
looks like unrelated changes?

russellb commented on src/instructlab/utils.py:
```diff
@@ -27,7 +27,6 @@
 
 DEFAULT_YAML_RULES = """\
 extends: relaxed
-
```
looks like an unrelated accidental change here

cdoern replied:
added back

russellb commented on src/instructlab/lab.py:
```diff
@@ -47,20 +46,22 @@
     config.init_config(ctx, config_file)
 
 
+ilab.add_command(model_group.model)
+ilab.add_command(taxonomy_group.taxonomy)
+ilab.add_command(data_group.data)
+ilab.add_command(config_group.config)
```
Was every commit prior to this one broken without these changes?

I don't think it's worth blocking on at this point, but it would be easier to follow all the changes with each of these included with the commit that created the group

russellb replied:
It looks like we're merging a bunch of commits that we know are in a broken state prior to this one. That's really not ideal, but I'm OK not blocking on it this time.

russellb commented on src/instructlab/lab.py:
```diff
@@ -36,7 +35,52 @@
+    def get_command(self, ctx, cmd_name):
+        if cmd_name in self.aliases:
+            cmd = self.aliases[cmd_name]["cmd"]
+            print(
+                f"You are using an aliased command, this will be deprecated in a future release. Please consider using `ilab {group} {c}` instead"
+            )
```
Is it possible to use `click.echo()` here and color it in red?

I would also adjust your wording here to say that this command is deprecated now and will be removed in a future release vs will be deprecated in the future.

russellb replied:
Also what are `{group}` and `{c}` ? It looks like they're not defined

cdoern replied:
added back, I accidentally deleted these

russellb commented on tests/test_lab_download.py:
```diff
@@ -16,15 +16,15 @@ class TestLabDownload:
                 [
                     "--config=DEFAULT",
-                    "download",
+                    "model" "download",
```
This is missing a command between "model" and "download".

I'm assuming the test is failing. I did see the test CI job is failing

russellb commented on tests/test_lab_download.py:
```diff
@@ -43,7 +43,7 @@ def test_download_error(self):
                 [
                     "--config=DEFAULT",
-                    "download",
+                    "model" "download",
```
Same thing here -- missing a command between "model" and "download"

Hopefully it's showing up in the current test failures too

russellb reviewed (changes requested):
This is a huge improvement over the last time I read through this. Thank you for your diligence in working through the feedback!

I left some comments inline. I tried to highlight when it's something I wanted to point out but am willing to not block on it. Feel free to tag me for clarification wherever needed, though.

There's one thing from my past review I saw still outstanding, `chain=True`, which AFAICT shouldn't be needed.

From the click docs:

> Sometimes it is useful to be allowed to invoke more than one subcommand in one go. For instance if you have installed a setuptools package before you might be familiar with the setup.py sdist bdist_wheel upload command chain which invokes sdist before bdist_wheel before upload. 

I don't think that's the behavior we want, or at least I'd rather not introduce it in this PR as it would be a much bigger UX change than I think you intended.

Can you file issues to track updating:

- all docs for the command changes
- all tests

russellb commented on src/instructlab/lab.py:
```diff
@@ -541,788 +479,6 @@ def generate(
-)
-@utils.macos_requirement(echo_func=click.secho, exit_exception=click.exceptions.Exit)
-# pylint: disable=function-redefined
-def test(data_dir, model_dir, adapter_file):
```
... just kidding, i forgot that this was renamed to evaluate!

cdoern replied:
test --> evaluate. but ilab test still maps to it for now

russellb replied:
yep, sorry about that.  i tried to comment before you had to correct me on it haha

cdoern commented on src/instructlab/model/chat.py:
```diff
@@ -55,6 +62,209 @@
 PROMPT_PREFIX = ">>> "
 
 
+@click.command()
```
I fixed the context passing! I had to map the ctx.obj and the ctx.default_map to the ctx.parent's equivalent.

russellb replied:
nice work! It looks MUCH better now

cdoern commented on requirements.txt:
```diff
@@ -40,3 +40,4 @@ langchain-text-splitters
 # the below library should NOT be imported into any python files
 # it is for CLI usage ONLY
 yamllint>=1.35.1,<1.36.0
+prettytable==3.10.0 
```
in the unit test the highest version found by pip as 3.10.0. Unsure why but it seemed to have a lot of python version dependencies so I pinned it the the highest which all versions of python we build for support.

russellb replied:
Can you add a comment in the file that explains why we're stuck on this specific version?

alimaredia replied:
We can remove this since `ilab model list` was removed from the PR

russellb replied:
I think this change is worth making before merging

cdoern commented on src/instructlab/model/serve.py:
```diff
@@ -0,0 +1,70 @@
+@click.option(
+    "--model-family",
+    type=str,
+    default="merlinite",
```
fixed, thank you for catching this

alimaredia commented on pyproject.toml:
```diff
@@ -109,18 +109,33 @@ known-local-folder = ["tuning"]
     "^src/instructlab/train/lora_mlx/utils\\.py$",
     "^src/instructlab/utils\\.py$",
+    "^src/instructlab/server\\.py$",
+    "^src/instructlab/helpers\\.py$",
```
what file is `helpers` pointing to?

cdoern replied:
old file, nice catch :pray:

russellb replied:
That file existed in previous versions of this PR< but is gone now

alimaredia commented on pyproject.toml:
```diff
@@ -109,18 +109,33 @@ known-local-folder = ["tuning"]
+    "^src/instructlab/model/list\\.py$",
+    "^src/instructlab/model/model\\.py$",
+    "^src/instructlab/model/train\\.py$",
+    "^src/instructlab/config/config\\.py$",
```
Shouldn't this be `"^src/instructlab/configuration/config\\.py$",`

alimaredia commented on pyproject.toml:
```diff
@@ -109,18 +109,33 @@ known-local-folder = ["tuning"]
+    "^src/instructlab/model/model\\.py$",
+    "^src/instructlab/model/train\\.py$",
+    "^src/instructlab/config/config\\.py$",
+    "^src/instructlab/config/init\\.py$",
```
ditto

cdoern replied:
nice catch, thanks

cdoern replied:
(it was once called config)

cdoern commented on src/instructlab/config.py:
```diff
@@ -14,15 +17,12 @@
     ValidationError,
     field_validator,
 )
+import click
```
moved them

xukai92 commented on src/instructlab/lab.py:
```diff
@@ -146,7 +146,7 @@ def init(
                 return
         click.echo(
-            "Welcome to InstructLab CLI. This guide will help you to setup your environment."
+            "Welcome to InstructLab ilab. This guide will help you to setup your environment."
```
this looks like an accident change to me as a result of replacing all `cli` with `ilab`.
if so please revert.

alimaredia commented on src/instructlab/model/chat.py:
```diff
@@ -18,11 +20,16 @@
+# First Party
+from instructlab import client, config
+from instructlab.config import DEFAULT_CONNECTION_TIMEOUT, DEFAULT_MODEL_OLD
+from instructlab.server import is_temp_server_running
```
This is minor but this could be moved into the function level import in `def chat()` since it's only used there

alimaredia commented on src/instructlab/model/chat.py:
```diff
@@ -18,11 +20,16 @@
 
+# First Party
+from instructlab import client, config
+from instructlab.config import DEFAULT_CONNECTION_TIMEOUT, DEFAULT_MODEL_OLD
```
This is minor but this could be moved into the function level import in def chat_cli() since it's only used there

xukai92 commented on src/instructlab/configuration/config.py:
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
```
i'm a bit confused by what the commit message means

russellb replied:
same ... but not going to block on it

xukai92 reviewed (commented):
just a quick question about having something like `model/model.py` v.s. having the code in `model/__init__.py` -- which one is a better practice?

cdoern commented:
my recent set of commits:

1. added --config back to top level `ilab` command by manually passing ctx down in each level by retrieving objects from `ctx.parent.```
2. fixed some small nits from reviews, removed some unnecessary things here and there
3. renamed tax to taxonomy

alimaredia commented on .gitignore:
```diff
@@ -3,6 +3,7 @@ models/
 !src/instructlab/train/lora_mlx/models
 # We recommend using taxonomy here, but not a submodule. Just a work dir.
 taxonomy/
+!src/instructlab/taxonomy/
```
what is the point in adding this to the `.gitignore`?

cdoern replied:
the new dir is called `taxonomy` for the taxonomy group

alimaredia replied:
Got that, what are we expecting to go into that directory that developers would want to ignore?

alimaredia commented on src/instructlab/taxonomy/taxonomy.py:
```diff
@@ -0,0 +1,25 @@
+def taxonomy(ctx):
+    """Command Group for Interacting with the Taxonomy of InstructLab.
+
+    If this is your first time running ilab, it's best to start with `ilab init` to create the environment.
```
As was mentioned in a comment in `src/instructlab/data/data.py`. it should be `ilab config init` not `ilab init`.

russellb replied:
https://github.com/instructlab/instructlab/issues/1331

alimaredia commented on src/instructlab/config.py:
```diff
@@ -216,3 +223,50 @@ def get_model_family(forced, model_path):
+        elif not os.path.isfile(config_file):
+            config_obj = None
+            ctx.fail(
+                f"`{config_file}` does not exists, please run `ilab init` "
```
Shouldn't this say please run `ilab config init`?

russellb replied:
... and should say "does not exist" instead of "exists"

but I don't think we need to block on this. It's pre-existing code and `ilab init` still works

alimaredia replied:
https://github.com/instructlab/instructlab/issues/1331

alimaredia commented on src/instructlab/lab.py:
```diff
@@ -79,1296 +95,21 @@ def __init__(self, config_obj: config.Config):
-    Returns a torch.device object:
-    - type is one of 'cpu', 'cuda', 'hpu'
-    - index is None or device index (e.g. 0 for first GPU)
+    If this is your first time running ilab, it's best to start with `ilab init` to create the environment.
```
Another spot where it should say `ilab config init` instead of `ilab init`

russellb replied:
https://github.com/instructlab/instructlab/issues/1331

alimaredia reviewed (dismissed):
@cdoern this PR looks great, and I think it's ready to merge, with LOTs of cleanup after. 
I think at this point I'd rather merge the PR as is, and distribute fixing minor cleanups to the community as opposed to putting it all on Charlie to keep having to do.

That being said, any/all cleanup should be meticulously tracked and fixed before the next release. I know this is tedious, but we can't just forget about it.

There are just 2 things I wanted to discuss/clarify.

1. I think we should get rid of `evaluate.py` and the `ilab model evaluate` command. This is because we know we're going to drastically change the functionality of the command as it is now. Also since we were going to replace it with something else we should either get rid of what it's doing entirely OR change it to `ilab model test`. I think we should get rid of it entirely because we would just have a follow-up PR removing `ilab model test` anyway. @russellb what do you think.

If the PR is merged before we can do this, please create an issue for this. 

2. Out of curiousity, what was the logic behind the files in `tests` that were changed? Did you want to have some tests that tested the new functionality but still leave the scripts behind to test backwards compat?

leseb commented:
> @cdoern this PR looks great, and I think it's ready to merge, with LOTs of cleanup after. I think at this point I'd rather merge the PR as is, and distribute fixing minor cleanups to the community as opposed to putting it all on Charlie to keep having to do.
> 
> That being said, any/all cleanup should be meticulously tracked and fixed before the next release. I know this is tedious, but we can't just forget about it.
> 
> There are just 2 things I wanted to discuss/clarify.
> 
> 1. I think we should get rid of `evaluate.py` and the `ilab model evaluate` command. This is because we know we're going to drastically change the functionality of the command as it is now. Also since we were going to replace it with something else we should either get rid of what it's doing entirely OR change it to `ilab model test`. I think we should get rid of it entirely because we would just have a follow-up PR removing `ilab model test` anyway. @russellb what do you think.

I would like to do this in a follow-up PR, so it's clear what happened (and when) to this file. We should only focus on the new structure and not change any existing things. Thanks!

> 
> If the PR is merged before we can do this, please create an issue for this.
> 
> 2. Out of curiousity, what was the logic behind the files in `tests` that were changed? Did you want to have some tests that tested the new functionality but still leave the scripts behind to test backwards compat?

cdoern commented:
sure, I can rename evaluate back to test. After my next push I am going to begin making follow up issues. 

In my next commits I:

1. removed the `chain=True` for each group
2. added DYM to each resource group
3. switched evaluate back to test

russellb commented on src/instructlab/data/data.py:
```diff
@@ -0,0 +1,32 @@
+def data(ctx, config_file):
+    """Command Group for Interacting with the Data generated by InstructLab.
+
+    If this is your first time running ilab, it's best to start with `ilab init` to create the environment.
```
https://github.com/instructlab/instructlab/issues/1331

russellb commented on src/instructlab/configuration/config.py:
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
```
https://github.com/instructlab/instructlab/issues/1332

russellb reviewed (changes requested):
There are now a bunch of follow-up issues open. 

The ONLY thing I'd like to request changed at this point before merging is dropping the new `prettytable` dependency since it is no longer used.

cdoern commented:
thanks for catching the prettytable dependency @russellb , this should be all set now

russellb reviewed (approved):
You have shown an impressive level of diligence as you worked through feedback. Thank you for sticking through this and pushing so hard to the end. Very nice work!

I'm good with merging this now!

alimaredia commented:
@leseb @markstur If your changes requested are still relevant can you point them out or if not remove the changes requested. It's the only thing standing in the way of merging the PR.

russellb commented:
I tried to apply `hold` to catch the merge before it went in but just missed it.

I talked with @cdoern about the merge process and how we have Mergify configured to explicitly NOT merge when someone has previously requested changes and not come back and explicitly ACKed that they have been resolved. My preference would have been to ping both @markstur and @leseb to give them a chance to give that ACK. Only if they didn't respond would I consider dismissing the review.

I know @cdoern had all the best of intentions here and genuinely felt that it was acceptable to dismiss the reviews if the feedback was addressed. In his defense, this is not documented anywhere! As I think about it, that's probably fine when the feedback in the review was simple enough. @leseb at least was doing more in depth review and has been very responsive on this.

@cdoern has offered to revert the PR to give a chance for the final review responses. I told him I did not feel it was worth the churn in this case.

@markstur @leseb would you both be OK taking another look at this PR (post-merge) to see if there was anything else you'd like to see addressed? If it's something not captured in a follow-up issue, please raise it. If you have any more significant concerns, we can consider reverting whatever portion is affected, but I'd rather hold off on that churn unless it's justified.

Hopefully that makes sense to everyone! Let me know if I can help discuss anything further.

markstur commented:
> I tried to apply `hold` to catch the merge before it went in but just missed it.
> 
> I talked with @cdoern about the merge process and how we have Mergify configured to explicitly NOT merge when someone has previously requested changes and not come back and explicitly ACKed that they have been resolved. My preference would have been to ping both @markstur and @leseb to give them a chance to give that ACK. Only if they didn't respond would I consider dismissing the review.
> 
> I know @cdoern had all the best of intentions here and genuinely felt that it was acceptable to dismiss the reviews if the feedback was addressed. In his defense, this is not documented anywhere! As I think about it, that's probably fine when the feedback in the review was simple enough. @leseb at least was doing more in depth review and has been very responsive on this.
> 
> @cdoern has offered to revert the PR to give a chance for the final review responses. I told him I did not feel it was worth the churn in this case.
> 
> @markstur @leseb would you both be OK taking another look at this PR (post-merge) to see if there was anything else you'd like to see addressed? If it's something not captured in a follow-up issue, please raise it. If you have any more significant concerns, we can consider reverting whatever portion is affected, but I'd rather hold off on that churn unless it's justified.
> 
> Hopefully that makes sense to everyone! Let me know if I can help discuss anything further.

yeah I was trying to ACK those when it merged.  Now I'm looking at main.  So far the concerns seem to be fixed or no longer relevant in what was merged

cdoern commented:
@markstur @leseb I am really sorry for not letting you guys ACK the PR before merging it, it was an honest mistake as I genuinely thought stale reviews which I had either made follow up issues for or addressed in the PR, could be dismissed. I did not realize this was not a normal practice and will be careful doing it in the future. Please feel free to be honest with me and let me know if you'd like this reverted, I really apologize.

markstur commented:
> @markstur @leseb I am really sorry for not letting you guys ACK the PR before merging it, it was an honest mistake as I genuinely thought stale reviews which I had either made follow up issues for or addressed in the PR, could be dismissed. I did not realize this was not a normal practice and will be careful doing it in the future. Please feel free to be honest with me and let me know if you'd like this reverted, I really apologize.

@cdoern  No worries.  No revert needed for any of mine.  Just double-checking to make sure they are resolved or no longer relevant in main.  So far, so good.

leseb commented:
> @markstur @leseb I am really sorry for not letting you guys ACK the PR before merging it, it was an honest mistake as I genuinely thought stale reviews which I had either made follow up issues for or addressed in the PR, could be dismissed. I did not realize this was not a normal practice and will be careful doing it in the future. Please feel free to be honest with me and let me know if you'd like this reverted, I really apologize.

@cdoern No worries, I'm not offended by how things have unfolded. There's no need to revert; we have follow-ups in place, and any missed items will be addressed.
